google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java

```diff
@@ -356,6 +400,65 @@ public void run(Throwable finalStatus) {
     log.info("Finish connecting stream: " + streamName + " id: " + writerId);
   }
 
+  @GuardedBy("lock")
+  private boolean shouldWaitForBackoff(AppendRequestAndResponse requestWrapper) {
+    if (this.retrySettings != null
+        && Instant.now().isBefore(requestWrapper.blockMessageSendDeadline)) {
+      log.fine(
+          String.format(
+              "Waiting for waiting to queue to unblock at %s for retry # %s",
+              requestWrapper.blockMessageSendDeadline, requestWrapper.retryCount));
+      return true;
+    }
+
+    return false;
+  }
+
+  private void waitForBackoffIfNecessary(AppendRequestAndResponse requestWrapper) {
+    lock.lock();
+    try {
+      Condition condition = lock.newCondition();
+      while (shouldWaitForBackoff(requestWrapper)) {
+        condition.await(100, java.util.concurrent.TimeUnit.MILLISECONDS);
+      }
+    } catch (InterruptedException e) {
+      throw new IllegalStateException(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private boolean isImmediatelyRetriableStatus(Code statusCode) {
```

Comment:
I think I'd prefer to keep them separate, as they have different purposes (connection error retry vs immediate in-stream retry).  I could also see a world where they diverge.

Reply:
Since the code in each is identical, I suggest placing the functions next to each other. For now we can simply have the existing function isConnectionErrorRetriable() call the new function isImmediatelyRetriableStatus(). That way if they need to diverge in future the change could be made in isConnectionErrorRetriable().

Reply:
I ended up combining them; if we need to split it out later it's not a big deal.